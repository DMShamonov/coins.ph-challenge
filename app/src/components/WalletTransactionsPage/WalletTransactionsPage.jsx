// @flow

import * as React from 'react';

import {
  Page,
  PageHeader,
  PageContent,
  PageTitle,
} from 'components/Page';
import Flex from 'components/Flex';
import Total from 'components/Total';
import WalletTransactions from 'components/WalletTransactions';
import Button from 'components/Button';
import WalletTransactionCreateModal from 'components/WalletTransactionCreateModal';

import type { PropsType } from './types.js.flow';

export default function WalletTransactionsPage({
  wallet,
  walletTransactions,
  onOpenCreateWalletTransactionModal,
}: PropsType): React.Node {
  return (
    <React.Fragment>
      <Page className="wallets-page">
        <PageHeader>
          <Flex>
            <PageTitle>{`"${wallet.label}" transactions`}</PageTitle>
            <Total value={walletTransactions.length} />
          </Flex>
          <Button onClick={onOpenCreateWalletTransactionModal}>Create transaction</Button>
        </PageHeader>
        <PageContent>
          {
            walletTransactions.length
              ? <WalletTransactions walletId={wallet.id} walletTransactions={walletTransactions} />
              : 'You haven\'t wallet transactions yet...'}
        </PageContent>
      </Page>
      <WalletTransactionCreateModal walletId={wallet.id} />
    </React.Fragment>
  );
}
