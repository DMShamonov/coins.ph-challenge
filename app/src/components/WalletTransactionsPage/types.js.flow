// @flow

import type { WalletType, WalletTransactionsType } from 'types.js.flow';

export type PropsType = {
  wallet: WalletType,
  walletTransactions: WalletTransactionsType,
  onOpenCreateWalletTransactionModal: () => void,
};
