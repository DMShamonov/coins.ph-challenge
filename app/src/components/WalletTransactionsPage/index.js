import { compose } from 'redux';
import { connect } from 'react-redux';

import WALLETS_ACTION_TYPES from 'redux/actionTypes/wallets';
import WALLET_TRANSACTIONS_ACTION_TYPES from 'redux/actionTypes/walletTransactions';
import SESSION_ACTION_TYPES from 'redux/actionTypes/session';
import MODAL_ACTION_TYPES from 'redux/actionTypes/modal';
import MODALS from 'constants/modals';
import preloader from 'decorators/preloader';
import api from 'controllers/api';
import { comparator } from 'controllers/utilities';

import WalletTransactionsPage from './WalletTransactionsPage';

function mapStateToProps({ data: { wallets, walletTransactions } }, { match: { params: { id } } }) {
  return {
    wallet: wallets.find(wallet => wallet.id === id),
    walletTransactions: walletTransactions.sort(
      ({ date: dateA }, { date: dateB }) => comparator(dateA, dateB, true),
    ),
  };
}

function mapDispatchToProps(dispatch) {
  return {
    onOpenCreateWalletTransactionModal() {
      dispatch({
        type: MODAL_ACTION_TYPES.ON_SHOW,
        payload: { modal: MODALS.CREATE_WALLET_TRANSACTION },
      });
    },
    dispatch,
  };
}

export default compose(
  connect(
    mapStateToProps,
    mapDispatchToProps,
  ),
  preloader(async ({ match: { params: { id } }, wallet: walletData, dispatch }) => {
    try {
      const wallet = await api().wallets().get({ id });
      const { transactions } = await wallet.transactions();

      if (!walletData) {
        dispatch({ type: WALLETS_ACTION_TYPES.ON_ADD, payload: { wallets: [wallet.wallet] } });
      }

      dispatch({ type: WALLET_TRANSACTIONS_ACTION_TYPES.ON_SET, payload: { transactions } });
    } catch (e) {
      console.warn(e); // eslint-disable-line no-console

      if (e?.status === 401) {
        dispatch({ type: SESSION_ACTION_TYPES.LOGOUT });
      }
    }
  }),
)(WalletTransactionsPage);
