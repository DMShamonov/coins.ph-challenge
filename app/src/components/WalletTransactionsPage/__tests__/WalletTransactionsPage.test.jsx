import React from 'react';
import { shallow } from 'enzyme';

import WalletTransactionsPage from '../WalletTransactionsPage';

describe('`Wallet` component should render correctly', () => {
  test('with default props', () => {
    expect(shallow((
      <WalletTransactionsPage
        wallet={{ id: '__ID__', label: '__LABEL__', balance: 100000000 }}
        walletTransactions={[{
          id: '__ID__',
          date: '2019-04-09T02:56:04.152Z',
          entries: [{ account: '__WALLET_ID__', value: 100000000 }],
          fee: 1000,
          pending: false,
        }]}
      />
    ))).toMatchSnapshot();
  });

  test('when `walletTransactions` prop is empty array', () => {
    expect(shallow((
      <WalletTransactionsPage
        wallet={{ id: '__ID__', label: '__LABEL__', balance: 100000000 }}
        walletTransactions={[]}
      />
    ))).toMatchSnapshot();
  });
});
