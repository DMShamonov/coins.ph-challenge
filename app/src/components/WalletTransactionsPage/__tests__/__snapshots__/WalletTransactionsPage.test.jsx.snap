// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`Wallet\` component should render correctly when \`walletTransactions\` prop is empty array 1`] = `
<Fragment>
  <Page
    className="wallets-page"
  >
    <PageHeader>
      <Flex>
        <PageTitle>
          "__LABEL__" transactions
        </PageTitle>
        <Total
          value={0}
        />
      </Flex>
      <Button>
        Create transaction
      </Button>
    </PageHeader>
    <PageContent>
      You haven't wallet transactions yet...
    </PageContent>
  </Page>
  <Connect(WalletTransactionCreateModal)
    walletId="__ID__"
  />
</Fragment>
`;

exports[`\`Wallet\` component should render correctly with default props 1`] = `
<Fragment>
  <Page
    className="wallets-page"
  >
    <PageHeader>
      <Flex>
        <PageTitle>
          "__LABEL__" transactions
        </PageTitle>
        <Total
          value={1}
        />
      </Flex>
      <Button>
        Create transaction
      </Button>
    </PageHeader>
    <PageContent>
      <WalletTransactions
        walletId="__ID__"
        walletTransactions={
          Array [
            Object {
              "date": "2019-04-09T02:56:04.152Z",
              "entries": Array [
                Object {
                  "account": "__WALLET_ID__",
                  "value": 100000000,
                },
              ],
              "fee": 1000,
              "id": "__ID__",
              "pending": false,
            },
          ]
        }
      />
    </PageContent>
  </Page>
  <Connect(WalletTransactionCreateModal)
    walletId="__ID__"
  />
</Fragment>
`;
